pkgs/http/lib/src/io_client.dart

```diff
@@ -123,7 +120,76 @@ class IOClient extends BaseClient {
         ioRequest.headers.set(name, value);
       });
 
-      var response = await stream.pipe(ioRequest) as HttpClientResponse;
+      // SDK request aborting is only effective up until the request is closed,
+      // at which point the full response always becomes available.
+      // This occurs at `pipe`, which automatically closes the request once the
+      // request stream has been pumped in.
+      // Therefore, we have multiple strategies:
+      //  * If the user aborts before we have a response, we can use SDK abort,
+      //    which causes the `pipe` (and therefore this method) to throw the
+      //    aborted error
+      //  * If the user aborts after we have a response but before they listen
+      //    to it, we immediately emit the aborted error then close the response
+      //    as soon as they listen to it
+      //  * If the user aborts whilst streaming the response, we inject the
+      //    aborted error, then close the response
+
+      var isAborted = false;
+      var hasResponse = false;
+
+      if (request case Abortable(:final abortTrigger?)) {
+        unawaited(
+          abortTrigger.whenComplete(() {
+            isAborted = true;
+            if (!hasResponse) ioRequest.abort(RequestAborted(request.url));
+          }),
+        );
+      }
+
+      final response = await stream.pipe(ioRequest) as HttpClientResponse;
+      hasResponse = true;
+
+      StreamSubscription<List<int>>? ioResponseSubscription;
+
+      late final StreamController<List<int>> responseController;
+      responseController = StreamController(
+        onListen: () {
+          if (isAborted) {
+            responseController
+              ..addError(RequestAborted(request.url))
+              ..close();
+            return;
+          } else if (request case Abortable(:final abortTrigger?)) {
+            abortTrigger.whenComplete(() {
+              if (!responseController.isClosed) {
+                responseController
+                  ..addError(RequestAborted(request.url))
+                  ..close();
+              }
+              ioResponseSubscription?.cancel();
+            });
+          }
+
+          ioResponseSubscription = response.listen(
```

Comment:
Do you think that it would be worth the reduction in complexity to write this as:

```
ioResponseSubscription = response.listen(...
if (isAborted) {
  responseController
  ..addError(RequestAborted(request.url))
  ..close();
  ioResponseSubscription.cancel();
}

Reply:
I don't necessarily see how this would make anything easier. Surely anything not in the `onListen` callback would require another asynchronous trigger/completer, which is more complex. Still need to wait for a listener & the abort trigger.

I think having the separation of concerns with the two trigger listeners makes sense to me.